File: springboot/graphql/src/main/java/com/graphql/demo/article/Article.java
package com.graphql.demo.article;

import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.DynamicUpdate;

import javax.persistence.*;
import java.time.LocalDateTime;

@NoArgsConstructor
@Getter
@Entity
@DynamicUpdate
public class Article {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long idx;

    @Column(nullable = false)
    private String subject;

    @Column(nullable = false)
    private String contents;

    private LocalDateTime createdDate;

    @Builder
    private Article(String subject, String contents) {
        this.subject = subject;
        this.contents = contents;
        this.createdDate = LocalDateTime.now();
    }
}

File: springboot/HibernateLock/src/main/kotlin/com/example/hibernatelock/article/entity/Article.kt
package com.example.hibernatelock.article.entity

import javax.persistence.*


@Entity
@Table(name = "article3")
class Article(
    subject: String,
) {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var idx: Long? = null
        protected set

    @Column(nullable = false)
    var subject: String = subject
        protected set

    var viewCount: Long = 0
        protected set

    @Version
    var version: Int = 0
        protected set

    fun updateContent(subject: String) {

        this.subject = subject
    }
}